2_process/src/lake_sequences_mntoha.py

```diff
@@ -0,0 +1,202 @@
+import numpy as np
+import pandas as pd
+
+def all_dates_depths(lake_obs_interpolated, depths, depth_col='interpolated_depth', pad_before_days=0, pad_after_days=0, min_days=1):
+    """
+    Create a DataFrame with one column per depth and one row for every day from
+    `pad_before_days` days before the first observation until `pad_after_days`
+    after the last observation. Missing observations are filled in with np.NaN.
+
+    :param lake_obs_interpolated: DataFrame of observations for one lake, interpolated to the depths in the list `depths`.
+    :param depths: Full list of depths at which temperatures will be provided by the trained model.
+    :param depth_col: Name of DataFrame column with depth values to match up with `depths`. (Default value = 'interpolated_depth')
+    :param pad_before_days: Number of days to pad before the first observation (Default value = 0)
+    :param pad_after_days: Number of days to pad after the last observation (Default value = 0)
+    :param min_days: Minimum number of days included in output DataFrame (Default value = 1)
+    :returns: DataFrame of observations with NaN values for missing observations.
+
+    """
+
+    pad_before_days_td = np.timedelta64(pad_before_days, 'D')
+    pad_after_days_td = np.timedelta64(pad_after_days, 'D')
+    min_days_td = np.timedelta64(min_days, 'D')
+    one_day = np.timedelta64(1, 'D')
+    # buffer start_date by pad_before_days days
+    start_date = lake_obs_interpolated.date.min() - pad_before_days_td
+    # buffer end_date by pad_after_days days
+    end_date = lake_obs_interpolated.date.max() + pad_after_days_td
+    # check if enough days are included, pad if not
+    if end_date - start_date + one_day < min_days_td:
+        start_date = end_date - min_days_td + one_day
+    date_range = pd.date_range(start=start_date, end=end_date)
+
+    # create MultiIndex of all combinations of dates and depths
+    midx = pd.MultiIndex.from_product([date_range, depths], names=['date', depth_col])
+    # Average duplicate measurements (same day, same nearest depth)
+    obs_midx = lake_obs_interpolated.groupby(['date', depth_col]).temp.mean().to_frame()
+    obs_all = obs_midx.reindex(midx, fill_value=np.NaN)
+
+    # reshape to have one column per depth
+    obs_all = obs_all.unstack()
+    # flatten column levels
+    obs_all.columns = [' '.join([str(s) for s in col]).strip() for col in obs_all.columns.values]
```

Comment:
Phew took me a bit to unpack what was happening here - had to go mock it up in a notebook. I'd forgotten that `unstack()` generates those weird column levels. I think this could be written more simply by iterating through both elements of the column names:
```python
obs_all.columns = [f'{i} {j}' for i,j in obs_all.columns.values]
```
And I'd suggest adding a note in your comment that `unstack()` generates two column levels: `temp` and the interpolated depths, and it's those values that are being flattened here to 'temp 0.0', etc. Also - this is nitpicking, esp. b/c the column names get dropped later when you convert the dataframe to a numpy array, but typically we try to avoid spaces in column names and use underscores instead.

Reply:
Thanks for that snippet - so much nicer! I went through `all_dates_depths` and rewrote the comments to improve readability.

Reply:
Thanks Andy - these new comments are super helpful